marshal.go

```diff
@@ -99,112 +136,156 @@ const rxBufSize = 65535 // max size of IPv4 & IPv6 packet
-				err = fmt.Errorf("Request timeout (after %d retries)", retries-1)
+			if x.OnRetry != nil {
+				x.OnRetry(x)
+			}
+
+			x.Logger.Printf("Retry number %d. Last error was: %v", retries, err)
+			if withContextDeadline && strings.Contains(err.Error(), "timeout") {
+				err = context.DeadlineExceeded
 				break
 			}
 			if retries > x.Retries {
-				// Report last error
+				if strings.Contains(err.Error(), "timeout") {
+					err = fmt.Errorf("request timeout (after %d retries)", retries-1)
+				}
 				break
 			}
+			if x.ExponentialTimeout {
+				// https://www.webnms.com/snmp/help/snmpapi/snmpv3/v1/timeout.html
+				timeout *= 2
+			}
+			withContextDeadline = false
 		}
 		err = nil
 
-		reqDeadline := time.Now().Add(x.Timeout / time.Duration(x.Retries+1))
-		x.Conn.SetDeadline(reqDeadline)
+		if x.Context.Err() != nil {
+			return nil, x.Context.Err()
+		}
+
+		reqDeadline := time.Now().Add(timeout)
+		if contextDeadline, ok := x.Context.Deadline(); ok {
+			if contextDeadline.Before(reqDeadline) {
+				reqDeadline = contextDeadline
+				withContextDeadline = true
+			}
+		}
+
+		err = x.Conn.SetDeadline(reqDeadline)
+		if err != nil {
+			return nil, err
+		}
 
-		// Request ID is an atomic counter (started at a random value)
-		reqID := atomic.AddUint32(&(x.requestID), 1) // TODO: fix overflows
+		// Request ID is an atomic counter that wraps to 0 at max int32.
+		reqID := (atomic.AddUint32(&(x.requestID), 1) & 0x7FFFFFFF)
 		allReqIDs = append(allReqIDs, reqID)
 
 		packetOut.RequestID = reqID
 
 		if x.Version == Version3 {
-			msgID := atomic.AddUint32(&(x.msgID), 1) // TODO: fix overflows
-			allMsgIDs = append(allMsgIDs, msgID)
+			msgID := (atomic.AddUint32(&(x.msgID), 1) & 0x7FFFFFFF)
+
+			// allMsgIDs = append(allMsgIDs, msgID) // unused
 
 			packetOut.MsgID = msgID
 
 			err = x.initPacket(packetOut)
 			if err != nil {
 				break
 			}
-
 		}
-		if x.loggingEnabled && x.Version == Version3 {
+		if x.Version == Version3 {
 			packetOut.SecurityParameters.Log()
 		}
 
 		var outBuf []byte
 		outBuf, err = packetOut.marshalMsg()
 		if err != nil {
 			// Don't retry - not going to get any better!
-			err = fmt.Errorf("marshal: %v", err)
+			err = fmt.Errorf("marshal: %w", err)
 			break
 		}
 
-		dst, err := net.ResolveUDPAddr("udp", x.Target + ":" + strconv.Itoa(int(x.Port)))
-		if err != nil {
-			x.logPrintf("ERROR on ResolveUDPAddr: %s", err)
-			err = fmt.Errorf("Error on ResolveUDPAddr: %s", err.Error())
-			continue
+		if x.PreSend != nil {
+			x.PreSend(x)
+		}
+		x.Logger.Printf("SENDING PACKET: %#+v", *packetOut)
+		// If using UDP and unconnected socket, send packet directly to stored address.
+		if uconn, ok := x.Conn.(net.PacketConn); ok && x.uaddr != nil {
+			_, err = uconn.WriteTo(outBuf, x.uaddr)
+		} else {
+			_, err = x.Conn.Write(outBuf)
 		}
-		x.logPrintf("SENDING PACKET: %#+v", *packetOut)
-		_, err = x.Conn.WriteTo(outBuf, dst)
 		if err != nil {
 			x.logPrintf("ERROR on Conn.WriteTo: %s", err)
```

Comment:
![image](https://user-images.githubusercontent.com/5491690/186058380-1d92799d-10fb-4a5a-a111-12175f0a9acc.png)

![image](https://user-images.githubusercontent.com/5491690/186058409-f0c83154-3f25-4c1d-ab5a-995098651276.png)

```suggestion
			x.Logger.Printf("ERROR on Conn.WriteTo: %s", err)
```

Reply:
I'm  not sure if that happens on my fork or not. My GoLand makes everything red so I can't see errors amongst the sea of blood